### DISCUSSION TITLE
Didn't find PN53x board

### DISCUSSION THREAD

**Discussion by electric-m** [rednblkx/HomeKey-ESP32#35] [Q&A] [answered]
Hey there,

I did successfully setup one device a few weeks ago.
Today, I tried to setup a a second device with the exact same code.

Everything works fine, except, that the PN532 reader is not recognized:

```
Device Name:      HK SchlupfSchloss

I (2849) gpio: GPIO[5]| InputEn: 1| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (2854) gpio: GPIO[18]| InputEn: 1| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (2863) gpio: GPIO[19]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (2872) gpio: GPIO[23]| InputEn: 1| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
E (2888) NFC_SETUP: Didn't find PN53x board
Configuring LockManagement
I (2889) LockMechanism: LockMechanism > Configuring LockMechanism
I (2899) NFCAccess: NFCAccess > Configuring NFCAccess
```

I do remember, that I had this error with the first device as well, but do not remember, how I got away with it. It works like a charm today.
Any hints from you guys?

**Comment by electric-m** [comment]
Well, I found my own post #32 ,but that does not seem to be the issue actually. Different power supplies still don't help to recognize the reader.

**Comment by electric-m** [comment]
Thats the build output: (Something wrong here?)
```
 *  Das Terminal wird von Aufgaben wiederverwendet, drücken Sie zum Schließen eine beliebige Taste. 

 *  Task wird ausgeführt: platformio run 

Processing release (platform: espressif32; board: wemos_d1_mini32; framework: arduino, espidf)
------------------------------------------------------------------------------------------------------------------------------------------------------------
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/espressif32/wemos_d1_mini32.html
PLATFORM: Espressif 32 (6.5.0) > WEMOS D1 MINI ESP32
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
DEBUG: Current (cmsis-dap) External (cmsis-dap, esp-bridge, esp-prog, iot-bus-jtag, jlink, minimodule, olimex-arm-usb-ocd, olimex-arm-usb-ocd-h, olimex-arm-usb-tiny-h, olimex-jtag-tiny, tumpa)
PACKAGES: 
 - framework-arduinoespressif32 @ 3.20014.231204 (2.0.14) 
 - framework-espidf @ 3.40406.240122 (4.4.6) 
 - tool-cmake @ 3.16.4 
 - tool-esptoolpy @ 1.40501.0 (4.5.1) 
 - tool-ninja @ 1.9.0 
 - toolchain-esp32ulp @ 1.23500.220830 (2.35.0) 
 - toolchain-xtensa-esp32 @ 8.4.0+2021r2-patch5
Warning! Arduino framework as an ESP-IDF component doesn't handle the `variant` field! The default `esp32` variant will be used.
Reading CMake configuration...
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ deep, Compatibility ~ soft
Found 38 compatible libraries
Scanning dependencies...
Dependency Graph
|-- HomeSpan @ 1.8.0+sha.e21a6b2
|-- PN532 @ 1.0.0+sha.ab92bff
|-- YACL @ 1.0.3
|-- nlohmann-json @ 3.11.3
|-- PicoMQTT @ 0.3.8+sha.640a332
|-- SPI @ 2.0.0
Building in release mode
Compiling .pio/build/release/src/main.o
src/main.cpp: In member function 'virtual void LockMechanism::loop()':
src/main.cpp:164:12: warning: unused variable 'exchange' [-Wunused-variable]
       bool exchange = nfc.inDataExchange(data, sizeof(data), selectCmdRes, &selectCmdResLength);
            ^~~~~~~~
src/main.cpp: In member function 'virtual boolean NFCAccess::update(std::vector<char>*, int*)':
src/main.cpp:335:11: warning: unused variable 'dataConfState' [-Wunused-variable]
     char* dataConfState = configurationState->getNewString();
           ^~~~~~~~~~~~~
src/main.cpp: In member function 'int NFCAccess::set_reader_key(uint8_t*, size_t)':
src/main.cpp:316:11: warning: 'void* memcpy(void*, const void*, size_t)' pointer overflow between offset 0 and size [4294967295, 2147483647] [-Warray-bounds]
     memcpy(readerData.reader_private_key, readerKey, tlv8.len(kReader_Req_Reader_Private_Key));
     ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/main.cpp:316:11: warning: 'void* memcpy(void*, const void*, size_t)' specified size 4294967295 exceeds maximum object size 2147483647 [-Wstringop-overflow=]
Linking .pio/build/release/firmware.elf
Retrieving maximum program size .pio/build/release/firmware.elf
Checking size .pio/build/release/firmware.elf
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [==        ]  16.7% (used 54864 bytes from 327680 bytes)
Flash: [========= ]  94.3% (used 1483969 bytes from 1572864 bytes)
Building .pio/build/release/firmware.bin
esptool.py v4.5.1
Creating esp32 image...
Merged 26 ELF sections
Successfully created esp32 image.
=============================================================== [SUCCESS] Took 16.62 seconds ===============================================================

Environment    Status    Duration
-------------  --------  ------------
release        SUCCESS   00:00:16.616
=============================================================== 1 succeeded in 00:00:16.616 ===============================================================
 *  Das Terminal wird von Aufgaben wiederverwendet, drücken Sie zum Schließen eine beliebige Taste.
 * ```

**Comment by electric-m** [comment]
More:
```
Compiling .pio/build/release/lib154/PN532/PN532_HSU.cpp.o
In file included from .pio/libdeps/release/PN532/src/PN532.cpp:11:
.pio/libdeps/release/PN532/src/PN532_debug.h:11: warning: "SERIAL" redefined
     #define SERIAL Serial
 
In file included from .pio/libdeps/release/PN532/src/PN532.cpp:9:
/Users/michael/.platformio/packages/framework-arduinoespressif32/cores/esp32/Arduino.h:50: note: this is the location of the previous definition
 #define SERIAL  0x0
 
In file included from .pio/libdeps/release/PN532/src/PN532_HSU.cpp:3:
.pio/libdeps/release/PN532/src/PN532_debug.h:11: warning: "SERIAL" redefined
     #define SERIAL Serial
 
In file included from .pio/libdeps/release/PN532/src/PN532_HSU.h:6,
                 from .pio/libdeps/release/PN532/src/PN532_HSU.cpp:2:
/Users/michael/.platformio/packages/framework-arduinoespressif32/cores/esp32/Arduino.h:50: note: this is the location of the previous definition
 #define SERIAL  0x0
 
Compiling .pio/build/release/lib154/PN532/PN532_I2C.cpp.o
Compiling .pio/build/release/lib154/PN532/PN532_SPI.cpp.o
.pio/libdeps/release/PN532/src/PN532.cpp: In member function 'uint32_t PN532::writeRegister(uint16_t, uint8_t)':
.pio/libdeps/release/PN532/src/PN532.cpp:185:14: warning: unused variable 'response' [-Wunused-variable]
     uint32_t response;
              ^~~~~~~~
Compiling .pio/build/release/lib154/PN532/emulatetag.cpp.o
In file included from .pio/libdeps/release/PN532/src/PN532_I2C.cpp:6:
.pio/libdeps/release/PN532/src/PN532_debug.h:11: warning: "SERIAL" redefined
     #define SERIAL Serial
 
In file included from .pio/libdeps/release/PN532/src/PN532_debug.h:6,
                 from .pio/libdeps/release/PN532/src/PN532_I2C.cpp:6:
/Users/michael/.platformio/packages/framework-arduinoespressif32/cores/esp32/Arduino.h:50: note: this is the location of the previous definition
 #define SERIAL  0x0
 
Archiving .pio/build/release/libcdd/libWire.a
Indexing .pio/build/release/libcdd/libWire.a
In file included from .pio/libdeps/release/PN532/src/PN532_SPI.cpp:3:
.pio/libdeps/release/PN532/src/PN532_debug.h:11: warning: "SERIAL" redefined
     #define SERIAL Serial
 
In file included from .pio/libdeps/release/PN532/src/PN532_debug.h:6,
                 from .pio/libdeps/release/PN532/src/PN532_SPI.cpp:3:
/Users/michael/.platformio/packages/framework-arduinoespressif32/cores/esp32/Arduino.h:50: note: this is the location of the previous definition
 #define SERIAL  0x0
 
Compiling .pio/build/release/lib154/PN532/llcp.cpp.o
.pio/libdeps/release/PN532/src/PN532_HSU.cpp: In member function 'virtual void PN532_HSU::wakeup()':
.pio/libdeps/release/PN532/src/PN532_HSU.cpp:31:17: warning: unused variable 'ret' [-Wunused-variable]
         uint8_t ret = _serial->read();
                 ^~~
.pio/libdeps/release/PN532/src/PN532_HSU.cpp: In member function 'virtual int8_t PN532_HSU::writeCommand(const uint8_t*, uint8_t, const uint8_t*, uint8_t)':
.pio/libdeps/release/PN532/src/PN532_HSU.cpp:46:17: warning: unused variable 'ret' [-Wunused-variable]
         uint8_t ret = _serial->read();
                 ^~~
Archiving .pio/build/release/lib0be/libSPI.a
In file included from .pio/libdeps/release/PN532/src/emulatetag.cpp:10:
.pio/libdeps/release/PN532/src/PN532_debug.h:11: warning: "SERIAL" redefined
     #define SERIAL Serial
 
In file included from .pio/libdeps/release/PN532/src/PN532_debug.h:6,
                 from .pio/libdeps/release/PN532/src/emulatetag.cpp:10:
/Users/michael/.platformio/packages/framework-arduinoespressif32/cores/esp32/Arduino.h:50: note: this is the location of the previous definition
 #define SERIAL  0x0
 
Indexing .pio/build/release/lib0be/libSPI.a
Compiling .pio/build/release/lib154/PN532/mac_link.cpp.o
In file included from .pio/libdeps/release/PN532/src/llcp.cpp:3:
.pio/libdeps/release/PN532/src/PN532_debug.h:11: warning: "SERIAL" redefined
     #define SERIAL Serial
 
In file included from .pio/libdeps/release/PN532/src/PN532_debug.h:6,
                 from .pio/libdeps/release/PN532/src/llcp.cpp:3:
/Users/michael/.platformio/packages/framework-arduinoespressif32/cores/esp32/Arduino.h:50: note: this is the location of the previous definition
 #define SERIAL  0x0
 
In file included from .pio/libdeps/release/PN532/src/mac_link.cpp:3:
.pio/libdeps/release/PN532/src/PN532_debug.h:11: warning: "SERIAL" redefined
     #define SERIAL Serial
 
In file included from .pio/libdeps/release/PN532/src/PN532_debug.h:6,
                 from .pio/libdeps/release/PN532/src/mac_link.cpp:3:
/Users/michael/.platformio/packages/framework-arduinoespressif32/cores/esp32/Arduino.h:50: note: this is the location of the previous definition
 #define SERIAL  0x0
 
.pio/libdeps/release/PN532/src/PN532_SPI.cpp: In member function 'virtual int16_t PN532_SPI::readResponse(uint8_t*, uint8_t, uint16_t)':
.pio/libdeps/release/PN532/src/PN532_SPI.cpp:183:17: warning: unused variable 'POSTAMBLE' [-Wunused-variable]
         uint8_t POSTAMBLE = read(); // POSTAMBLE
                 ^~~~~~~~~
Compiling .pio/build/release/lib154/PN532/snep.cpp.o
.pio/libdeps/release/PN532/src/emulatetag.cpp: In member function 'bool EmulateTag::emulate(uint16_t)':
.pio/libdeps/release/PN532/src/emulatetag.cpp:135:12: warning: unused variable 'cc_size' [-Wunused-variable]
   uint16_t cc_size = sizeof(compatibility_container);
            ^~~~~~~
Compiling .pio/build/release/liba57/YACL/CBOR.cpp.o
In file included from .pio/libdeps/release/PN532/src/snep.cpp:3:
.pio/libdeps/release/PN532/src/PN532_debug.h:11: warning: "SERIAL" redefined
     #define SERIAL Serial
 
In file included from .pio/libdeps/release/PN532/src/PN532_debug.h:6,
                 from .pio/libdeps/release/PN532/src/snep.cpp:3:
/Users/michael/.platformio/packages/framework-arduinoespressif32/cores/esp32/Arduino.h:50: note: this is the location of the previous definition
 #define SERIAL  0x0
 
Compiling .pio/build/release/liba57/YACL/CBORArray.cpp.o
Compiling .pio/build/release/liba57/YACL/CBORPair.cpp.o
.pio/libdeps/release/YACL/src/CBOR.cpp: In member function 'bool CBOR::add(char)':
.pio/libdeps/release/YACL/src/CBOR.cpp:526:12: warning: comparison is always false due to limited range of data type [-Wtype-limits]
  if (value < 0) {
```

**Comment by rednblkx** [comment]
Hi!
I can confirm nothing changed on the code side recently in relation to the nfc module. Since the code compiles successfully, issue is not with the code.

1. Make sure you used the following pinout:
  GPIO18 - SCK
  GPIO19 - MISO
  GPIO23 - MOSI
  GPIO5 - SS
2. Check that you set the DIP switch on the nfc module to SPI if it has one
3. Use different wires to account for faulty wires

It’s easy to overlook those little things so it is sometimes good to give a bit more attention to the simple things.

Can’t really think of anything else, hopefully someone else might have a better idea.

**Comment by electric-m** [comment]
> 1. Make sure you used the following pinout:
>    GPIO18 - SCK
>    GPIO19 - MISO
>    GPIO23 - MOSI
>    GPIO5 - SS
> 2. Check that you set the DIP switch on the nfc module to SPI if it has one
> 3. Use different wires to account for faulty wires
> 
> It’s easy to overlook those little things so it is sometimes good to give a bit more attention to the simple things.

I totally agree with that. I checked all these points multiple times, including completes set of new wires.

I am in no way a code expert, what made me wonder were those warnings while compiling.
They were related to the the PN532 so I thought it might be a problem.

```
Compiling .pio/build/release/lib154/PN532/snep.cpp.o
.pio/libdeps/release/PN532/src/emulatetag.cpp: In member function 'bool EmulateTag::emulate(uint16_t)':
.pio/libdeps/release/PN532/src/emulatetag.cpp:135:12: warning: unused variable 'cc_size' [-Wunused-variable]
   uint16_t cc_size = sizeof(compatibility_container);
            ^~~~~~~
            

```

Just one example from above.

**Reply by rednblkx** [reply]
Well… like I said before, nothing changed on that regard. Since the code worked before it should also work now….meaning we are definitely missing something 😶‍🌫️.
And no need to worry about the warnings, they are merely a compliance issue not an actual issue.
If you can also try using another module, chances are there might something wrong with that one.

**Reply by electric-m** [reply]
Right. Out of desperation I desoldered and redid **all** of the connections and triple checked, that they were good.
I measured and checked the power supply.
Now, I am definitely out of ideas and giving up on it at least for the moment, since I don't have a spare PN532.

**Comment by electric-m** [comment]
I just tried to whole process with a brand new esp23 Wemos D1 mini, still no success. So I suppose it's the PN532 board.

**Reply by 011V32** [reply]
Did you get it working?
I have the same problem with an brand new setup. I checked the connections for continuity.
![image](https://github.com/rednblkx/HomeKey-ESP32/assets/22346481/bd8d6eb3-dd23-4ac3-acc4-dc3ffd2c1baf)
![image](https://github.com/rednblkx/HomeKey-ESP32/assets/22346481/7c04d5b9-32e4-4d9f-8bd6-0e77b338cc19)

**Reply by xrust83** [reply]
VCC GND no? :-)

**Reply by ihrapsa** [reply]
Don't forget the VCC and GND pins. The PN532 module needs that to get power and have same voltage reference as the ESP.

**Reply by 011V32** [reply]
That's what I initially asked myself, but since only the cable with 4 wires was included and only the 4 cables were described in the wiki, I did it that way.
[NFC Module Wiring](https://github.com/rednblkx/HomeKey-ESP32/wiki/Prerequisites#nfc-module-wiring)
Thx.

**Reply by rednblkx** [reply]
Did not realize this might not be general knowledge, it is now included on the wiki, thank you.

**Reply by 011V32** [reply]
Sry, but I have another question. In the description of the PN532 board, it states that the operating voltage is 3.3V, but the VCC pin (which you added now in the wiki) has 5V. (The 3V3 pin would have 3.3V)

**Reply by rednblkx** [reply]
The PN532 takes in both 3.3V and 5V and the latter would be more preferred for stability.

**Reply by 011V32** [reply]
OK. Then I'll try my luck with 5V.
In the description I found the following: "On-board level shifter, standard 5V TTL for I2C and UART, 3.3V TTL for SPI"

**Reply by rednblkx** [reply]
yes, as it says it has an on-board level shifter, this translates 5V to the 3.3V logic level.

**Comment by electric-m** [comment]
I just want to add, that in my case, the power supply is not missing. Today, I bought and used completely new PN532 board and D1 Mini (identical to the one I already got up and running).
Still, the board is not found.
I am pretty sure I wired everything correctly. Even measured all cables and connections for positive throughput. It has got to be software from this point on, right?
Since its probably not an issue with the code, it might be with Visual Studio Code/Platformio?
I am running out of ideas...

**Reply by 011V32** [reply]
Which branch do you use?

**Reply by electric-m** [reply]
See next post, it’s solved.

**Comment by electric-m** [comment, marked as answer]
Holy moly, I just solved this❗️
The issue was:
I had the D1 Mini connected to the Mac via USB for, obviously, uploading the code.
I left it connected and had the PN532 powered through another, separate power supply, and not the VCC/GND pins on the D1 mini as they never worked for me.
I now cut an old usb cable and connected it to the same power supply that the PN532 is powered of.
And, voila, here we go.
I can't access the Serial monitor that way, but I checked for the MQTT messages, and the all come through now, when holding a tag in front of the reader.
Probbaly both devices need the exact same power supply, or at least a common ground.

2 Things you guys might still have an idea on:

1. Why isn't it possible to power the the PN532 through the ESP pins? Or is this just me?
2. Any chance i can observe the Platformio serial console through the IP Adress of the ESP?

Happy hacking
Michael

**Reply by rednblkx** [reply]
Brilliant!
I’ve now read that you did mentioned other time you switched psu but for some reason didn’t struck me as odd at the time 🥸.
However, it makes perfect sense that would be the case since like @ihrapsa said, they need to share the same voltage reference, otherwise you end up with two closed circles and the nfc module would be generating signals based on it’s very own frequencies and the ESP likewise, guess that’s another one for the wiki. It’s not the most correct explanation but I’m also not an electrical engineer 🤪.

As to your 2 inquiries:
1. Could you detail exactly what is the issue that prevents you from powering the module from the ESP? That should totally be possible and I would know since I’ve done from the beginning. Don’t have a Mac, but, If I remember correctly they have/had an issue with delivering power through USB at least to some devices, you should try using a Powered USB Hub.
2. **Currently** that is not possible since nothing is broadcasting the logs but it is something that’s been on my list for sometime, generally that’s done using the EventSource protocol and is also how ESPHome does it.

**Reply by electric-m** [reply]
We’re getting theee, I can feel it.
Most certainly it’s just like you said.
In the first place, I had the Reader wired up to ESP completely, including VCC/GND.
But probably the Mac USB port didn’t deliver enough/stable power for the PN532 to be recognized correctly.
That is why I switched to the separate power supply, but stuck to the Mac USB port for seeing the serial console.
All messed up from there on, but it’s seems clear to me now.
I am a bit embarrassed, that I didn’t get to it sooner, because in very other odd project I‘ve always been using one single source of power for the exact same reason.

**Reply by rednblkx** [reply]
No worries, you had an issue, you found a solution even if it offspring another problem, part of the process. What matters is that you figured it out.

**Comment by ronnievanced** [comment]
Ho thnx , very good project! Do i really need MQTT ? I just dont wanna use smart home systems like home assistant. Or its necessary? I want to use lock only in my room door. 
2. Question : how i can use servo in this project? Where i need to add servo code ? 
Thnx.

**Reply by electric-m** [reply]
You might want to start a separate topic for that, as this one is about the PN532 not being recognized.
